device/common/pddf/plugins/sfputil.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python
+
+import os.path
+import sys, traceback
+sys.path.append('/usr/share/sonic/platform/plugins')
+import pddfparse
+import json
+
+try:
+    import time
+    from ctypes import create_string_buffer
+    from sonic_sfp.sfputilbase import SfpUtilBase
+except ImportError, e:
+    raise ImportError (str(e) + "- required module not found")
+
+class SfpUtil(SfpUtilBase):
+    """Platform generic PDDF SfpUtil class"""
+
+    _port_to_eeprom_mapping = {}
+    _port_start = 0
+    _port_end = 0
+    _port_to_type_mapping = {}
+    _qsfp_ports = []
+    _sfp_ports = []
+
+    def __init__(self):
+        SfpUtilBase.__init__(self)
+        global pddf_obj
+        global plugin_data
+        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)) + '/../pddf/pd-plugin.json')) as pd:
+            plugin_data = json.load(pd)
+
+        pddf_obj = pddfparse.PddfParse()
+        self.platform = pddf_obj.get_platform()
+        self._port_start = 0
+        self._port_end = self.get_num_ports()
+
+        for port_num in range(self._port_start, self._port_end):
+            device = "PORT" + "%d"%(port_num+1)
+            port_eeprom_path = pddf_obj.get_path(device,"eeprom")
+            self._port_to_eeprom_mapping[port_num] = port_eeprom_path
+            port_type = pddf_obj.get_device_type(device)
+            self._port_to_type_mapping[port_num] = port_type
+            ret = self.populate_port_type(port_num)
+
+    def get_num_ports(self):
+        return int(self.platform['num_ports'])
+
+    def is_valid_port(self, port_num):
+        if port_num < self._port_start or port_num > self._port_end:
+            return False
+        else:
+            return True
+
+    def get_presence(self, port_num):
+        if port_num < self._port_start or port_num > self._port_end:
+            return False
+
+        status = 0
+        device = "PORT" + "%d"%(port_num+1)
+        port_ps = pddf_obj.get_path(device,"xcvr_present")
+        #print "port_ps value is : %s"%port_ps
+        try:
+            reg_file = open(port_ps, 'r')
+        except IOError as e:
+            print "Error: unable to open file: %s" % str(e)
+            return False
+
+        reg_value = reg_file.readline().rstrip()
+        #print "%s: Presence Status %d"%(device, int(reg_value))
+        if reg_value == '1':
+            return True
+
+        return False
+
+    def populate_port_type(self, port):
+        if self._port_to_type_mapping[port] == 'QSFP':
+            self._qsfp_ports.append(port)
+        elif self._port_to_type_mapping[port] == 'SFP':
+            self._sfp_ports.append(port)
+
```

Comment:
These port types are supposed to added by ODM when a new platform support is created example SPF28 or DDQSFP ?